File: typescript/shared/format.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.printTasksTable = printTasksTable;
exports.printTasksCSV = printTasksCSV;
exports.printTasksJSON = printTasksJSON;
exports.printLabelsTable = printLabelsTable;
exports.printSprintReport = printSprintReport;
function printTasksTable(tasks) {
    const border = "+----+----------------------+--------+----------+------------+";
    const header = "| ID | Title                | Status | Priority | Labels     |";
    console.log(border);
    console.log(header);
    console.log(border);
    for (const t of tasks) {
        const labelsStr = t.labels.join(",");
        const status = t.status.toUpperCase();
        const priority = t.priority.toUpperCase();
        // Exact fixed padding matching borders
        const idField = t.id.toString().padEnd(2).slice(0, 2);
        const titleField = t.title.padEnd(20).slice(0, 20);
        const statusField = status.padEnd(6).slice(0, 6);
        const prioField = priority.padEnd(8).slice(0, 8);
        const labelsField = labelsStr.padEnd(10).slice(0, 10);
        console.log(`| ${idField} | ${titleField} | ${statusField} | ${prioField} | ${labelsField} |`);
    }
    console.log(border);
}
function printTasksCSV(tasks) {
    console.log("id,title,status,priority,labels");
    for (const t of tasks) {
        const labelsStr = t.labels.join(";");
        // Standard minimal quoting for title and labels if needed:
        const qTitle = `"${t.title.replace(/"/g, '""')}"`;
        const qLabels = `"${labelsStr.replace(/"/g, '""')}"`;
        console.log(`${t.id},${qTitle},${t.status},${t.priority},${qLabels}`);
    }
}
function printTasksJSON(tasks) {
    console.log(JSON.stringify(tasks));
}
function printLabelsTable(db) {
    const border = "+-------------+-------------+";
    const header = "| Label Name  | Task Count  |";
    console.log(border);
    console.log(header);
    console.log(border);
    const counts = {};
    for (const l of db.labels) {
        counts[l.name] = 0;
    }
    for (const t of db.tasks) {
        for (const l of t.labels) {
            if (counts[l] !== undefined) {
                counts[l]++;
            }
        }
    }
    for (const l of db.labels) {
        const nameField = l.name.padEnd(11).slice(0, 11);
        const countField = counts[l.name].toString().padEnd(11).slice(0, 11);
        console.log(`| ${nameField} | ${countField} |`);
    }
    console.log(border);
}
function printSprintReport(db) {
    const total = db.tasks.length;
    let completed = 0;
    let todo = 0;
    let doing = 0;
    let done = 0;
    let high = 0;
    let medium = 0;
    let low = 0;
    for (const t of db.tasks) {
        const status = t.status.toLowerCase();
        if (status === "todo") {
            todo++;
        }
        else if (status === "doing") {
            doing++;
        }
        else if (status === "done") {
            done++;
            completed++;
        }
        const prio = t.priority.toLowerCase();
        if (prio === "low") {
            low++;
        }
        else if (prio === "medium") {
            medium++;
        }
        else if (prio === "high") {
            high++;
        }
    }
    let percent = 0;
    if (total > 0) {
        percent = Math.floor((completed * 100) / total);
    }
    const progressBlocks = Math.floor(percent / 10);
    const blocksStr = "■".repeat(progressBlocks) + "□".repeat(10 - progressBlocks);
    console.log("========================================");
    console.log("          MURLI-WORK SPRINT REPORT      ");
    console.log("========================================");
    console.log(`Completion Rate : [${blocksStr}] ${percent}% (${completed}/${total} tasks)\n`);
    console.log("Status Breakdown:");
    console.log(`- TODO  : ${todo} tasks`);
    console.log(`- DOING : ${doing} tasks`);
    console.log(`- DONE  : ${done} tasks\n`);
    console.log("Priority Breakdown:");
    console.log(`- HIGH  : ${high} tasks`);
    console.log(`- MEDIUM: ${medium} tasks`);
    console.log(`- LOW   : ${low} tasks`);
    console.log("========================================");
}

File: rust/clap/src/format.rs
use crate::db::{Database, Task};
use std::collections::HashMap;

pub fn print_tasks_table(tasks: &[Task]) {
    let border = "+----+----------------------+--------+----------+------------+";
    let header = "| ID | Title                | Status | Priority | Labels     |";

    println!("{}", border);
    println!("{}", header);
    println!("{}", border);

    for t in tasks {
        let labels_str = t.labels.join(",");
        let status = t.status.to_uppercase();
        let priority = t.priority.to_uppercase();

        // Safe truncation to avoid panicked out-of-bounds slicing:
        let title_trunc = if t.title.len() > 20 { &t.title[..20] } else { &t.title };
        let status_trunc = if status.len() > 6 { &status[..6] } else { &status };
        let prio_trunc = if priority.len() > 8 { &priority[..8] } else { &priority };
        let labels_trunc = if labels_str.len() > 10 { &labels_str[..10] } else { &labels_str };

        println!(
            "| {:<2} | {:<20} | {:<6} | {:<8} | {:<10} |",
            t.id, title_trunc, status_trunc, prio_trunc, labels_trunc
        );
    }
    println!("{}", border);
}

pub fn print_tasks_csv(tasks: &[Task]) {
    println!("id,title,status,priority,labels");
    for t in tasks {
        let labels_str = t.labels.join(";");
        // Standard double quotes escaping:
        let q_title = format!("\"{}\"", t.title.replace('"', "\"\""));
        let q_labels = format!("\"{}\"", labels_str.replace('"', "\"\""));
        println!("{},{},{},{},{}", t.id, q_title, t.status, t.priority, q_labels);
    }
}

pub fn print_tasks_json(tasks: &[Task]) {
    if let Ok(data) = serde_json::to_string(tasks) {
        println!("{}", data);
    }
}

pub fn print_labels_table(db: &Database) {
    let border = "+-------------+-------------+";
    let header = "| Label Name  | Task Count  |";

    println!("{}", border);
    println!("{}", header);
    println!("{}", border);

    let mut counts = HashMap::new();
    for l in &db.labels {
        counts.insert(l.name.clone(), 0);
    }
    for t in &db.tasks {
        for l in &t.labels {
            if let Some(c) = counts.get_mut(l) {
                *c += 1;
            }
        }
    }

    for l in &db.labels {
        let name_trunc = if l.name.len() > 11 { &l.name[..11] } else { &l.name };
        let count = counts.get(&l.name).cloned().unwrap_or(0);
        println!("| {:<11} | {:<11} |", name_trunc, count);
    }
    println!("{}", border);
}

pub fn print_sprint_report(db: &Database) {
    let total = db.tasks.len();
    let mut completed = 0;
    let mut todo = 0;
    let mut doing = 0;
    let mut done = 0;

    let mut high = 0;
    let mut medium = 0;
    let mut low = 0;

    for t in &db.tasks {
        match t.status.to_lowercase().as_str() {
            "todo" => todo += 1,
            "doing" => doing += 1,
            "done" => {
                done += 1;
                completed += 1;
            }
            _ => {}
        }

        match t.priority.to_lowercase().as_str() {
            "low" => low += 1,
            "medium" => medium += 1,
            "high" => high += 1,
            _ => {}
        }
    }

    let percent = if total > 0 { (completed * 100) / total } else { 0 };
    let progress_blocks = percent / 10;
    
    let mut blocks_str = String::new();
    for i in 0..10 {
        if i < progress_blocks {
            blocks_str.push('■');
        } else {
            blocks_str.push('□');
        }
    }

    println!("========================================");
    println!("          MURLI-WORK SPRINT REPORT      ");
    println!("========================================");
    println!("Completion Rate : [{}] {}% ({}/{}/ tasks)\n", blocks_str, percent, completed, total);
    println!("Status Breakdown:");
    println!("- TODO  : {} tasks", todo);
    println!("- DOING : {} tasks", doing);
    println!("- DONE  : {} tasks\n", done);
    println!("Priority Breakdown:");
    println!("- HIGH  : {} tasks", high);
    println!("- MEDIUM: {} tasks", medium);
    println!("- LOW   : {} tasks", low);
    println!("========================================");
}
